runtime/lua/_textutils.lua

```diff
@@ -0,0 +1,164 @@
+-- Utility functions for rich text ftplugins (vim help, markdown)
+
+local M = {}
+
+--- Apply current colorscheme to lists of default highlight groups
+---
+--- Note: {patterns} is assumed to be sorted by occurrence in the file.
+--- @param patterns {start:string,stop:string,match:string}[]
+function M.colorize_hl_groups(patterns)
+  local ns = vim.api.nvim_create_namespace('nvim.vimhelp')
+  vim.api.nvim_buf_clear_namespace(0, ns, 0, -1)
+
+  local save_cursor = vim.fn.getcurpos()
+
+  for _, pat in pairs(patterns) do
+    local start_lnum = vim.fn.search(pat.start, 'c')
+    local end_lnum = vim.fn.search(pat.stop)
+    if start_lnum == 0 or end_lnum == 0 then
+      break
+    end
+
+    for lnum = start_lnum, end_lnum do
+      local word = vim.api.nvim_buf_get_lines(0, lnum - 1, lnum, true)[1]:match(pat.match)
+      if vim.fn.hlexists(word) ~= 0 then
+        vim.api.nvim_buf_set_extmark(0, ns, lnum - 1, 0, { end_col = #word, hl_group = word })
+      end
+    end
+  end
+
+  vim.fn.setpos('.', save_cursor)
+end
+
+local heading_queries = {
+  vimdoc = [[
+    (h1 (heading) @h1)
+    (h2 (heading) @h2)
+    (h3 (heading) @h3)
+    (column_heading (heading) @h4)
+  ]],
+  markdown = [[
+    (setext_heading
+      heading_content: (_) @h1
+      (setext_h1_underline))
+    (setext_heading
+      heading_content: (_) @h2
+      (setext_h2_underline))
+    (atx_heading
+      (atx_h1_marker)
+      heading_content: (_) @h1)
+    (atx_heading
+      (atx_h2_marker)
+      heading_content: (_) @h2)
+    (atx_heading
+      (atx_h3_marker)
+      heading_content: (_) @h3)
+    (atx_heading
+      (atx_h4_marker)
+      heading_content: (_) @h4)
+    (atx_heading
+      (atx_h5_marker)
+      heading_content: (_) @h5)
+    (atx_heading
+      (atx_h6_marker)
+      heading_content: (_) @h6)
+  ]],
+}
+
+local function hash_tick(bufnr)
+  return tostring(vim.b[bufnr].changedtick)
+end
+
+--- Extract headings from buffer
+--- @param bufnr integer buffer to extract headings from
+--- @return table[]? headings each entry has keys `bufnr`, `lnum`, `capture`, `text`
+local get_headings = vim.func._memoize(hash_tick, function(bufnr)
+  local lang = vim.treesitter.language.get_lang(vim.bo[bufnr].filetype)
+  if not lang then
```

Comment:
In which case would `lang` be nil?

Reply:
If there's no `filetype`.

Reply:
Right, I got that bit. But is that possible in a Markdown or Vimdoc file?

I ask this in case that this might silently ignore an error, and maybe `assert` would be more appropriate.

Reply:
Well, this is primarily intended to shut up luals, and short-circuiting in this case seemed the most reasonable approach.

But maybe returning an empty table is even more reasonable given the purpose here and simplifies the check later.